src/TestUtils.jl

```diff
@@ -0,0 +1,74 @@
+module TestUtils
+
+using LinearAlgebra
+using Random
+using Test
+
+using Distributions
+using LogExpFunctions: logistic
+
+using AbstractGPs
+using ApproximateGPs
+
+function generate_data()
+    X = range(0, 23.5; length=48)
+    # The random number generator changed in 1.6->1.7. The following vector was generated in Julia 1.6.
+    # The generating code below is only kept for illustrative purposes.
+    #! format: off
+    Y = [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0]
+    #! format: on
+    # Random.seed!(1)
+    # fs = @. 3 * sin(10 + 0.6X) + sin(0.1X) - 1
+    # # invlink = normcdf
+    # invlink = logistic
+    # ps = invlink.(fs)
+    # Y = @. rand(Bernoulli(ps))
+    return X, Y
+end
+
+dist_y_given_f(f) = Bernoulli(logistic(f))
+
+function build_latent_gp(theta)
+    variance = softplus(theta[1])
+    lengthscale = softplus(theta[2])
+    kernel = variance * with_lengthscale(SqExponentialKernel(), lengthscale)
+    return LatentGP(GP(kernel), dist_y_given_f, 1e-8)
+end
+
+function test_approximation_predictions(approx)
+    rng = MersenneTwister(123456)
+    N_cond = 5
+    N_a = 6
+    N_b = 7
+
+    # Specify prior.
+    f = GP(Matern32Kernel())
+    # Sample from prior.
+    x = collect(range(-1.0, 1.0; length=N_cond))
+    noise_scale = 0.1
+    fx = f(x, noise_scale^2)
+    y = rand(rng, fx)
+
+    jitter = 0.0  # not needed in Gaussian case
+    lf = LatentGP(f, f -> Normal(f, noise_scale), jitter)
+    f_approx_post = posterior(approx, lf(x), y)
+
+    @testset "AbstractGPs API" begin
+        a = collect(range(-1.2, 1.2; length=N_a))
+        b = randn(rng, N_b)
+        AbstractGPs.TestUtils.test_internal_abstractgps_interface(rng, f_approx_post, a, b)
+    end
+
+    @testset "exact GPR equivalence for Gaussian likelihood" begin
```

Comment:
Do we know for sure that a given approximation is going to be able to satisfy this? i.e. if we have a not-so-great approximation, is it possible that it'll not be able to satisfy this, but still satisfy our API? i.e. it could provide implementations of `posterior` and `approx_lml` which satisfy the API.

I'm all for having a collection of "performance of approximate inference" sanity checks that can be opted into, but I'm not sure we need them to be part of the API tests.

Am I missing anything here?

Reply:
I wouldn't want an approximation that isn't able to return the exact answer in the conjugate case :joy: I haven't come across any that would do that. But if it really bothers you we can do similar to the public vs internal api tests in AbstractGPs?

Reply:
or a flag that turns off this specific test... but I do think it should be there by default (someone who wants to write a bad approximation could then opt out of it)

Reply:
Note that it doesn't say "test API", it says "test approximation predictions". could rename it to "test approximation equivalence in gaussian case" though that's a bit of a mouthful...

Reply:
> I wouldn't want an approximation that isn't able to return the exact answer in the conjugate case 😂 

haha fair point.

> or a flag that turns off this specific test... but I do think it should be there by default (someone who wants to write a bad approximation could then opt out of it)

I think this kind of thing makes sense. Maybe just allow the user to specify the tolerance required for approximate equality?

> Note that it doesn't say "test API", it says "test approximation predictions". could rename it to "test approximation equivalence in gaussian case" though that's a bit of a mouthful...

Also a fair point. I just assumed we were talking about interface tests, rather than correctness tests. I think I'm happy with the current name given the context. A docstring detailing what this function does / does not test would be helpful though.